src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubectlBuildWrapper.java

```diff
@@ -98,126 +101,30 @@ public void setUp(Context context, Run<?, ?> build, FilePath workspace, Launcher
         context.env("KUBECONFIG", configFile.getRemote());
         context.setDisposer(new CleanupDisposer(tempFiles));
 
-        String tlsConfig;
-        if (caCertificate != null && !caCertificate.isEmpty()) {
-            FilePath caCrtFile = workspace.createTempFile("cert-auth", "crt");
-            String ca = caCertificate;
-            if (!ca.startsWith(BEGIN_CERTIFICATE)) {
-                ca = wrapWithMarker(BEGIN_CERTIFICATE, END_CERTIFICATE, ca);
-            }
-            caCrtFile.write(ca, null);
-            tempFiles.add(caCrtFile.getRemote());
-
-            tlsConfig = " --certificate-authority=" + caCrtFile.getRemote();
-        } else {
-            tlsConfig = " --insecure-skip-tls-verify=true";
+        final KubernetesAuth auth;
+        try {
+            auth = KubernetesAuthFactory.fromCredentialsId(credentialsId, serverUrl, null, true);
+        } catch (KubernetesAuthException e) {
+            throw new AbortException(e.getMessage());
         }
 
-        int status = launcher.launch()
-                .cmdAsSingleString("kubectl config --kubeconfig=\"" + configFile.getRemote()
-                        + "\" set-cluster k8s --server=" + serverUrl + tlsConfig)
-                .join();
-        if (status != 0) throw new IOException("Failed to run kubectl config "+status);
-
-        final StandardCredentials c = getCredentials();
-
-        String login;
-        if (c == null) {
+        if (auth == null) {
             throw new AbortException("No credentials defined to setup Kubernetes CLI");
         }
 
-        if (c instanceof FileCredentials) {
-            try (InputStream in = ((FileCredentials) c).getContent(); OutputStream out = configFile.write()) {
-                IOUtils.copy(in, out);
+        // create Kubeconfig
+        try (Writer w = new OutputStreamWriter(new FileOutputStream(configFile.getRemote()), "UTF-8")) {
+            if (auth instanceof KubernetesAuthKubeconfig) {
```

Comment:
With the change I suggest above, this block would become `w.write(auth.buildKubeConfig());`